Add episode 43 (#391)
trinodb/trino.io: _episodes/43.md

---
layout: episode
title: "43: Trino saves trips with Alluxio"
date: 2022-12-15
tags: cache alluxio orchestration storage file
youtube_id: "---Jp9d-1qQUH0---"
wistia_id: ""
sections:
   - title: "Concept of the episode"
     desc: "Data caching and orchestration"
     time: 0
   - title: "Concept of the episode"
     desc: "Trino without caching"
     time: 0
   - title: "Concept of the episode"
     desc: "Hive connector caching"
     time: 0
   - title: "Concept of the episode"
     desc: "What is Alluxio?"
     time: 0
   - title: "Concept of the episode"
     desc: "What is data orchestration?"
     time: 0
   - title: "Concept of the episode"
     desc: "Trino and Alluxio"
     time: 0
   - title: "PR of the episode"
     desc: "Alluxio/alluxio PR 13000 Add a doc for Trino"
     time: 0
   - title: "Demo of the episode"
     desc: "Running Trino on Alluxio"
     time: 0
---

## Hosts

* Brian Olsen, Developer Advocate at [Starburst](https://starburst.io)
  ([@bitsondatadev](https://twitter.com/bitsondatadev))
* Manfred Moser, Director of Information Engineering at 
  [Starburst](https://starburst.io)
  ([@simpligility](https://twitter.com/simpligility))

> ## Guests

Reviewer: empty line after

* Bin Fan, VP of Open Source at Alluxio and PMC maintainer of Alluxio open 
  source and TSC member of Presto ([@binfan](https://twitter.com/binfan))
* [Beinan Wang](https://www.linkedin.com/in/beinan/), Software Engineer at 
  Alluxio and Presto committer

## Concept of the episode: Data caching and orchestration

Out of all those petabytes of data you store, only a small fraction of it is
creating business value for you today. When you scan the same data multiple
times and transfer it over the wire, you're wasting time, compute cycles, and
ultimately money. This gets worse when you're pulling data across regions or
clouds from disaggregate Trino clusters. In situations like these, caching
solutions can make a tremendous impact on the latency and cost of your queries.

> ### Trino without caching

Reviewer: we need to add that this is about caching for object files.. and therefore only applies to object storage connectors (hive, iceberg, delta lake and hudi)

There seems to be a sizeable portion of the community who aren't using a
caching solution. Not all workloads will really benefit from caching. If you
are performing more writes than reads, the cache will need to constantly be
invalidated before performing each read. If you are scanning all your data to
run daily migrations, you would not benefit from caching. However, one of the

> most common use cases where Trino shines is interactive adhoc analysis. This 

Reviewer: analytics

type of querying is very fast in Trino, especially when using modern storage 
formats like Iceberg. 

### Hive connector caching

Trino has an [embedded caching engine](https://trino.io/docs/current/connector/hive-caching.html)
in the Hive connector. This is convenient as it ships with Trino, however, it 
does not work outside the Hive connector. The caching engine is 
[Rubix](https://github.com/qubole/rubix) which has not been developed over two
years. While this system works for simple Hive use cases, it fails to address

> use cases outside of Hive and hasn't been maintained since 2020. There many

Reviewer: There are

does not work outside the Hive connector. The caching engine is 
[Rubix](https://github.com/qubole/rubix). While this system works for simple
Hive use cases, it fails to address use cases outside of Hive and hasn't been
maintained since 2020. There are many features missing like security features
and support for more compute engines. 

### What is Alluxio?

Alluxio is world’s first open source data orchestration technology for analytics
and AI for the cloud. It provides a common interface enabling computation
frameworks to connect to numerous storage systems through a common interface.
Alluxio’s memory-first tiered architecture enables data access at speeds orders
of magnitude faster than existing solutions. Alluxio was originally developed at
Berkley Amp Labs, [and was originally called Tachyon](
https://amplab.cs.berkeley.edu/wp-content/uploads/2014/11/2014_socc_tachyon.pdf).
It was less focused on caching and data orchestration and more focused on
fault-tolerance via lineage and other techniques borrowed from Spark. 

Alluxio lies between data driven applications, such as Trino and Apache Spark,
and various persistent storage systems, such as Amazon S3, Google Cloud Storage,

> HDFS, Ceph, and Minio. Alluxio unifies the data stored in these different

Reviewer: isnt it MinIO ?

invalidated before performing each read. If you are scanning all your data to
run daily migrations, you would not benefit from caching. However, one of the
most common use cases where Trino shines is interactive adhoc analytics. This 
type of querying is very fast in Trino, especially when using modern storage 
formats like Iceberg.

### Two types of caching

There are two types of caching used with Trino. The first type caches the
results of a common query or sub query to optimize performance for any query
that overlaps with predicates to obtain the cached results. 

The other type of query is object file caching. Rather than store the results of
the query, you are caching the files from a file or object store that are
scanned as part of the query.

In this episode, we will focus on the latter type of caching. This will apply to
connectors like Hive, Iceberg, Delta Lake, and Hudi.

### Hive connector caching

Trino has an [embedded caching engine](https://trino.io/docs/current/connector/hive-caching.html)
in the Hive connector. This is convenient as it ships with Trino, however, it 
does not work outside the Hive connector. The caching engine is 

> [Rubix](https://github.com/qubole/rubix) which has not been developed over two

Reviewer: Rubix. 

remove the rest since it comes up in the ext sentence.

years. While this system works for simple Hive use cases, it fails to address
use cases outside of Hive and hasn't been maintained since 2020. There are many

> features missing like kubernetes and containerization, security features, 

Reviewer: remove "Kubernetes and containerization" and rework the rest as needed.

support for more compute engines. 

### What is Alluxio?

Alluxio is world’s first open source data orchestration technology for analytics
and AI for the cloud. It provides a common interface enabling computation
frameworks to connect to numerous storage systems through a common interface.
Alluxio’s memory-first tiered architecture enables data access at speeds orders
of magnitude faster than existing solutions. Alluxio was originally developed at

> Berkley Amp Labs, [formerly known as Tachyon](

Reviewer: Amp Labs, and was originally called Tachyon. It was less focussed ...

https://amplab.cs.berkeley.edu/wp-content/uploads/2014/11/2014_socc_tachyon.pdf),
and was less focused on caching and data orchestration and more focused on
fault-tolerance via lineage and other techniques borrowed from Spark. 

Alluxio lies between data driven applications, such as Trino and Apache Spark,
and various persistent storage systems, such as Amazon S3, Google Cloud Storage,
HDFS, Ceph, and Minio. Alluxio unifies the data stored in these different

>  storage systems, presenting unified client APIs and a global namespace to its

Reviewer: remove space in from of "storage systems.

upper layer data driven applications.

<p align="center">
<img align="center" width="75%" height="100%" src="/assets/episode/43/alluxio-architecture.png"/>
</p>

Alluxio is commonly used as a distributed shared caching service so compute
engines talking to Alluxio can transparently cache frequently accessed data,
especially from remote locations, to provide in-memory I/O throughput. Alluxio

> also makes using Object storage much simpler as you will not hit the storage

Reviewer: object

Reviewer: also .. this sentence feels wrong ... it doesnt really make things simpler does it?

Author: I'll avoid using "simpler" but rather specify that "they enable unifying all data sources to a single file namespace". This makes things simpler in some aspects if you have multiple object storages across different storage technologies, object storage in different regions, or stored across different clouds.

in the cloud if the file is already cached by Alluxio. Alluxio also handles the

> complex mapping client interface requests to underlying storage interfaces.

Reviewer: mapping of client

but also .. I dont understand what this means.

Author: Going to reword [with this comment in mind](https://github.com/trinodb/trino.io/pull/391#discussion_r1058481809).

<p align="center">
<img align="center" width="75%" height="100%" src="/assets/episode/43/inside-alluxio.png"/>
</p>

Source: <https://docs.alluxio.io/os/user/stable/en/Overview.html>

### What is data orchestration?

A data orchestration platform abstracts data access across storage systems,
virtualizes all the data, and presents the data via standardized APIs with
global namespace to data-driven applications. In the meantime, it should have
caching functionality to enable fast access to warm data. In summary, a data

> orchestration platform provides data-driven applications Data Accessibility,

Reviewer: get rid of the title case for all these ..

Data Locality, and Data Elasticity.

Source: <https://www.alluxio.io/blog/data-orchestration-the-missing-piece-in-the-data-world/>

### Trino and Alluxio: Expedia use case

Expedia needed to have the ability to query cross cluster over different regions
while simplifying the interface to their local data sources.

<p align="center">
<img align="center" width="100%" height="100%" src="/assets/episode/43/expedia-trino-alluxio.png"/>
</p>

> Source: [Unifying Cross-region Access in the Cloud at Expedia Group — The Path Toward Data Mesh in the Brand World](https://www.alluxio.io/blog/unifying-cross-region-access-in-the-cloud-at-expedia-group-the-path-toward-data-mesh-in-the-brand-world/)

Reviewer: I know its a source.. but get rid of the title case anyway .. its not readable here

## PR of the episode: Alluxio/alluxio PR 13000 Add a doc for Trino

[This episode's PR](https://github.com/Alluxio/alluxio/pull/13000) is actually

> one of those not located in the Trino repository. This PR comes from the Alluxio

Reviewer: ```suggestion
not located in a Trino repository. This PR comes from the Alluxio
```

repository. It happened in the wake of the rebranding from Presto to Trino. PRs
like this helped continue the Trino community as it grew awareness around the
new name, as well as, fixed any potential issues that occurred with the hasty
renaming we had to do.

This was submitted by Alluxio engineer, [David Zhu](https://github.com/yuzhu).
A huge thanks to David and his contributions to Trino as well!

## Demo of the episode: Running Trino on Alluxio

> This demo of the episode, covers how to configure Alluxio to use write through

Reviewer: write-through
